pkg/storegateway/indexheader/reader_pool.go

```diff
@@ -51,21 +55,83 @@ type ReaderPool struct {
 	lazyReaders   map[*LazyBinaryReader]struct{}
 }
 
-// NewReaderPool makes a new ReaderPool and starts a background task for unloading idle Readers if enabled.
-func NewReaderPool(logger log.Logger, lazyReaderEnabled bool, lazyReaderIdleTimeout time.Duration, metrics *ReaderPoolMetrics) *ReaderPool {
+type HeadersLazyLoaded struct {
+	Path   string
+	UserID string
+}
+
+type HeadersLazyLoadedTrackerState struct {
+	LazyLoadedBlocks map[string]int64 `json:"lazy_loaded_blocks"`
```

Comment:
If we're expecting the keys to be ULIDs, should we use the `ulid.ULID` type here rather than `string`? (Ditto for the value - is there a type we can use that communicates it should be a timestamp?)